### PR TITLE
Implement loading DDS textures at run-time (reverted)

### LINKED ISSUE #5748: Implement loading DDS textures at run-time
### Describe the project you are working on

An alternate editor for existing train simulator (not written in Godot).

### Describe the problem or limitation you are having in your project

The original project is using DDS format for textures. DDS format in Godot is handled as a texture loader only (during importing assets within the editor), but my tool must work with original data files. These files are external for Godot.

Godot can't handle DDS files via `Image.load()`. The method is returing `ERR_FILE_UNRECOGNIZED` at runtime. DDS is handled only at import stage.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

The proposal is about adding `ImageLoaderDDS` class, moving DDS handling code inside it, and reusing image loader (or part of it) in texture loader.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Here is a prototype version of a patch (currently only for 3.x branch):
https://github.com/marcinn/godot/tree/image-loader-dds

It allows to use DDS files directly:
```gdscript
var im = Image.new()
im.load("/path/to/external/texture.dds")
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It cannot be worked around by few lines of script. Image loading is a core of the engine.

### Is there a reason why this should be core and not an add-on in the asset library?

1. The patch seems to be fully compatible
2. It makes Godot more compliant with the table of supported formats: https://docs.godotengine.org/en/stable/tutorials/assets_pipeline/importing_images.html

### PR DESCRIPTION
**Pull request by marcinn**
Closes godotengine/godot-proposals#5748

Adds:
* `Image::load_dds_from_buffer()`
* ability to load DDS textures at runtime via `Image::load_from_file`

The DDS loader implementation is moved from texture loader to image loader.

Backport for 3.x is in progress and will be available soon.

### CODE REVIEW

**Comment by fire** [comment]
I think this is good to review for merge. Hopefully someone can help me out to review. 👍

File: modules/dds/image_loader_dds.cpp
Comment (marcinn) [inline modules/dds/image_loader_dds.cpp:101]: I'm wondering if this function name should start with an underscore. Even in the future it shouldn't conflict with anything else, but maybe it would be better for the convention. Or, because it is directly bound to `Image::_dds_mem_loader_func`, maybe it should be just named `_dds_mem_loader_func`?

**Comment by Calinou** [comment]
I've built a release export template without and with this PR, and this appears to be make no difference in binary size at all (not even 1 byte).

**Comment by marcinn** [comment]
I've compiled template_release and exported my test project. Seems to work fine (dds texture is loaded at runtime).
Could you compare sha of the binaries?

**Comment by Calinou** [comment]
> I've compiled template_release and exported my test project. Seems to work fine (dds texture is loaded at runtime). Could you compare sha of the binaries?

```
33ca6fbaaed1b38af0395ccd3c28d983e558e48acb2c81875e7d1b8ce1303b98  bin/godot.linuxbsd.template_release.x86_64 (this PR)
c3852aa853d7540ec0b56a325ca7ba729f82dec8efe2e437bf2a3f737dd0c4f8  bin/godot.linuxbsd.template_release.x86_64.vanilla
```

```
.rwxr-xr-x   65,502,000 hugo 24 Nov  2:03  godot.linuxbsd.template_release.x86_64 (this PR)
.rwxr-xr-x   65,502,000 hugo 24 Nov  2:03  godot.linuxbsd.template_release.x86_64.vanilla
```

**Comment by marcinn** [comment]
Thanks. 

> 65,502,000

Nicely rounded. 
Size of my binary is 87530672. We probably used different build parameters. 
I've compiled binary with `scons platform=linuxbsd tools=no target=template_release bits=64`

You can check if it works if you have some time. I'm attaching an example project. The icon path is set by default to `/tmp/icon.dds`. Without the icon file, or when dds at runtime does not work, you will see a white box. If dds at runtime works as expected, you will see a textured box.

[TestRuntimeDDS.zip](https://github.com/godotengine/godot/files/10080410/TestRuntimeDDS.zip)
[icon.zip](https://github.com/godotengine/godot/files/10080412/icon.zip)

File: modules/dds/image_loader_dds.cpp
Comment (marcinn) [reply modules/dds/image_loader_dds.cpp]: I've renamed this function to match the convention.

**Comment by akien-mga** [comment]
Amended to fix copyright headers for 2023+.

**Review by Calinou — APPROVED** [review]
Should be good to merge after rebasing.

**Comment by marcinn** [comment]
Rebased

**Review by mhilbrunner — APPROVED** [review]

**Comment by marcinn** [comment]
Thank you. Could this pr be merged now, please?

**Review by clayjohn — APPROVED** [review]
This looks correct to me. I have reviewed the code, but I haven't tested locally.

**Comment by YuriSizov** [comment]
@marcinn Could you please rebase? This looks well approved now, we can merge it if it's rebased.

**Comment by marcinn1** [comment]
Rebased

**Comment by YuriSizov** [comment]
Thanks!

**Comment by TokisanGames** [comment]
Since Godot 3, DDS files have been used natively, without conversion or import files. This is why we use DDS files and have over 800 in our game, because they do not get converted and are used exactly as we made them. Godot 4's DXT conversion from non-DDS files has bugs that we wish to avoid, and we want control over the exact format and algorithm used in DXT conversion.

This PR broke this (confirmed by bisecting). When I loaded up my game in 4.2-dev, it started converting all of my DDS files from a working DXT format to a broken DXT format. Importing didn't complete and hung Godot, requiring a forcequit. Files that have mipmaps now report there are none. Textures are grainy. I have a duplicate of all textures on my drive (as DDS and ctex), and I likely have artifacts and other issues with various textures as they have been converted to a different format I didn't ask for.

It causes issues such as this:
https://github.com/TokisanGames/Terrain3D/issues/190
 
I also got these error messages loading up the Terrain3d demo.
```
ERROR: Condition "is_compressed()" is true. Returning: USED_CHANNELS_RGBA
ERROR:ERROR:ERROR:   at: detect_used_channels (core/io/image.cpp:3325)
 Condition "is_compressed()" is true. Returning: USED_CHANNELS_RGBA
ERROR: Condition "is_compressed()" is true. Returning: USED_CHANNELS_RGBA
 Condition "is_compressed()" is true. Returning: USED_CHANNELS_RGBA
   at:  Condition "is_compressed()" is true. Returning: USED_CHANNELS_RGBA
   at:    at: detect_used_channels (core/io/image.cpp:3325)
   at: detect_used_channels (core/io/image.cpp:3325)
detect_used_channels (core/io/image.cpp:3325)
detect_used_channels (core/io/image.cpp:3325)
res://demo/textures/rock028_nrm_rgh.dds: Texture detected as used as a normal map in 3D. Enabling red-green texture com.
res://demo/textures/rock028_nrm_rgh.dds: Texture detected as used as a roughness map in 3D. Enabling roughness limiter .
res://demo/textures/rock028_nrm_rgh.dds: Texture detected as used in 3D. Enabling mipmap generation and setting the tex.
ERROR: Cannot generate mipmaps in compressed or custom image formats.
   at: (core/io/image.cpp:1820)
ERROR: Condition "is_compressed()" is true. Returning: USED_CHANNELS_RGBA
   at: detect_used_channels (core/io/image.cpp:3325)
```

Please revert this change so DDS files are not converted or modified unless specifically requested.

**Comment by clayjohn** [comment]
@TokisanGames That sounds very unexpected. This PR didn't touch the importer. So if your files were not importing before, they should not start being imported after this change

**Comment by akien-mga** [comment]
@clayjohn This PR does add ImageLoaderDDS, which handles all files with `.dds` extension.

**Comment by TokisanGames** [comment]
Unexpected or not, it was confirmed by bisecting from 4.1 to master. DDS files have never been imported or converted since I started using Godot in 2018. And nor have import files been generated. Now both is happening. Not only are import files generated, the contents of the import files show that my DDS files have been converted to a different format, which is not desired.

```
importer="texture"
type="CompressedTexture2D"
uid="uid://cvfy20jicfbmh"
path.s3tc="res://.godot/imported/rock030_alb_ht.png-54156133eddcafc657ad8fa12baffcfb.s3tc.ctex"
metadata={
"imported_formats": ["s3tc_bptc"],
"vram_texture": true
}
```

As for an MRP @marcinn, all you need is any DDS. Load it up in 4.1 and it won't import. Load it up in any after this commit and it will import/convert. If you want a more practical project, download Terrain3D and look at the demo and it will make all the terrain textures grainy.

**Comment by marcinn1** [comment]
> This PR does add ImageLoaderDDS, which handles all files with `.dds` extension.
> I can confirm that Godot is defaulting to importing dds textures now all of a sudden

That's look like implicite behaviour of Godot (unexpected in context of this PR).
Is possible to force default import settings per file type?

**Comment by clayjohn** [comment]
> Is possible to force default import settings per file type?

@marcinn1 It doesn't look like it. There is the added problem that DDS needs to be imported as a plain resource so that it is available in the export pck. So we can't just filter it out of import completely.

**Comment by marcinn1** [comment]
So this is the importer which not work as other importers, and it is required for export. Quite unclear / inconsistent.
The goal of this PR was to able load DDS at runtime, but not change importer settings. 

As far I understand:

1. The editor changed import behaviour because of registration of ".dds" extension for ImageLoader (!)
2. There is no possibility to force the editor to use "Keep file / No import" by default for particular resource type
3. DDS import was always buggy and wasn't used before ("Keep file / No import" was used by default)
4. I reused this buggy code in runtime DDS loader


Possible solutions:
* revert the PR, drop dds loading, and runtime and leave possibly broken code as it was before
* tell the Editor to use "Keep file / No import" by default for DDS
* or as a workaround: change DDS importer to pass-through/dummy/do-nothing implementation (if possible)

**Comment by clayjohn** [comment]
@TokisanGames Could you highlight what exactly the bugs you are facing with the current DXT compression? I'm not aware of any bugs currently. Ideally we would fix the compression bugs your team is facing so that you and others can use other image types other than DDS

**Comment by TokisanGames** [comment]
It's been a while since I looked, and it looks like some are fixed. The ones I had in mind were:

Always using DXT5, fixed https://github.com/godotengine/godot/issues/58012

Issues I've had with `Image::compress()` https://github.com/godotengine/godot/issues/79932
The key issue isn't the lack of the compression library on export, I've worked around that. The issue is the wrong auto selection of format and no control over the format I want. I have it working now, without issue, but had to go through a lot of discovery to get there. It's not a robust API.

Artifacts in reflections https://github.com/godotengine/godot/issues/57981
Maybe addressed with settings?

I typically do DDS conversion in Gimp and what I've found is even there the default algorithm doesn't look good on reflective surfaces. But they give a wide variety of options and better control over the conversion so I can find a compression algorithm that works well. So I'll maintain the practice of external conversion for that control. 

I also need the external tool for channel packing. We're going to build a channel packing tool into Terrain3D, and will probably have to write out PNG, which will be read in as DXT :/ so hopefully that will go smoothly.

**Comment by clayjohn** [comment]
Thanks for filling me in! I am very interested in improving our image compression workflow. While I can't promise that every barrier your team runs into will be removed, it really helps me prioritize stuff to know what is blocking you.

**Comment by TokisanGames** [comment]
Thanks for your efforts and interest. We're currently not blocked on image compression at all. We have workarounds or documented processes for everything. The dds change was an unpleasant surprise, but I'm glad @Calinou caught it before 4.2 was released. We are blocked on [cowdata limitations](https://github.com/godotengine/godot/pull/74582#issuecomment-1650082792) though.

**Pull request merged**
